### ISSUE TITLE
Unable to send json values true, false and null

### ISSUE THREAD

**Issue by bobifle** [RPTools/maptool#670]
MT 1.5.4
There's is no way to send the following json object:
```
{
  "key0": false,
  "key1": true,
  "key2": null
}
```

to the external world using Maptool REST apis.
Best scenario, you can send strings containings "true", "false", "null" but that may clash with the structure the api is expecting.

Proposed solution: json.set and json.append should jsonify the strings "false", "true", and "null" into the appropriate Java type so the JSON library properly construct the json object.

That way we gain the ability to send any JSON supported data type.
We loose the ability to properly encode the **actual strings** "false", "true", "null", however I humbly think it's still better that the inability to send any bool or null value.

**Comment by JamzTheMan** [comment]
Curious, has anyone tried using the new json.path.set/put commands to see what that returns?

**Comment by Phergus** [comment]
With the current code, if you pass in `true` or `false`, you get an error.

If you pass in "true" or "false" you just end up with the strings.

At least that's what MT shows.

**Comment by Merudo** [comment]
Maybe augment the REST functions to pass a `JSONObject` argument: 

{
  "asTRUE": "true",
  "asFALSE": "false",
  "asNULL": "null"
}

and the REST functions could automatically parse these values to `true`/`false`/`null`

**Comment by Azhrei** [comment]
I like @Merudo's suggestion as that both 9a) leaves it to the user to decide, and (b) makes it flexible enough that other strings could be used.

Item (b), if I'm reading his suggestion correctly, is saying that any string inside the JSON that is "true" would be converted to the boolean `true`; ditto as appropriate for "false" and "null".

Making this an optional parameter to the REST functions means that the user only needs to worry about it in that case, and only if the receiving system requires booleans in the first place.  If the JSON object or array contains the string "true" and it _shouldn't_ be converted, the user could specify a different mapping, such as `{ "asTRUE": "MyTrueValue" }` and then the string "MyTrueValue" would be converted to boolean `true` instead.

To make things easier, perhaps the default JSON object, shown in Merudo's post, could be used when an empty JSON object is passed?  So the user can just specify `{}` instead of having to type all that out?

**Comment by JamzTheMan** [comment]
Could work, I assume editing the parser to accept literal booleans is off the table? (It wouldn't be horrible if we could use true/false in place of 0/1 in other places?

**Comment by Azhrei** [comment]
The tokenizer already has problems (don't create a variable that starts with the letters `tru` or `fal` because you'll get errors).

Maybe the literal `true` and `false` already work in JSON objects?  Parsing is left up to the JSON library, after all.  But MTscript doesn't have a boolean type, so it might be a pain to use.

**Comment by JamzTheMan** [comment]
Is that a hard fix? (Adding true/false to MTscript parser) Asking for a friend. :-)

**Comment by Phergus** [comment]
Just looking at the PR code, the changes only modify the json.set function so that it does the conversion to the real `boolean` or `null` types.

Based on experimenting with the JSON data returned from Chummer (Shadowrun character generator), `json.get()` handles this already.

Example Chummer snippet:
```
{"$id": "10639",
"editionNumber": "5e",
"language": "de-de",
"siNnerMetaData": {
"$id": "10640",
"id": "009f0da8",
"visibility": {
  "$id": "10641",
  "id": null,
  "isPublic": false,
  "isGroupVisible": false,
  "userRights": []
}}}
```
```
[r: isPublic = json.get(json.get(json.get(test,"siNnerMetaData"),"visibility"),"isPublic")]
```
Returns the string "false".

**Comment by Azhrei** [comment]
Well, I looked into it many years ago and yeah, I think it was going to be hard.  I don't remember the details, though.  Maybe it just _looked_ hard but wouldn't turn out to be so?

If it does get fixed, it shouldn't break anything since variable names starting with those letters would currently fail anyway; the change should be upward compatible.

**Comment by bobifle** [comment]
As I mentioned in the bug description, it's currently assumed that you cannot send the strings "null", "true" and "false" anymore.

Merudo's proposal tries to fix the "what if I want to both send true and "true" ? ".

Once you implement, the additional parameter, what if the json object I'm sending contained the string I used "asTRUE" (json object can initialy come from the other side) do I need to check every values before selecting my new asTrue string value ? It never ends.

More importantly, I'm super relunctant to shift the burden of the mumbo jumbo I initiated onto the final user by adding a parameter to the MT (currently clean) rest apis. Not until it's absolutely necessary.

My humble opinon :o)

**Comment by Azhrei** [comment]
> Merudo's proposal tries to fix the "what if I want to both send true and "true" ? ".
> 
> Once you implement, the additional parameter, what if the json object I'm sending contained the string I used "asTRUE" (json object can initialy come from the other side) do I need to check every values before selecting my new asTrue string value ? It never ends.

Hm.  Logically, it's possible that people are already using the word `true` to represent that word and implementing this feature will break existing code.  Shouldn't we offer a way to mitigate that breakage?  Ideally, in a way that's relatively simple to implement?  (Which is why I suggested an empty JSON object could trigger some set of defaults that would restore previous behavior.)

Another option would be to add some kind of `rest.config()` function so that such configuration can be done once in an onCampaignLoad event.

> More importantly, I'm super relunctant to shift the burden of the mumbo jumbo I initiated onto the final user by adding a parameter to the MT (currently clean) rest apis. Not until it's absolutely necessary.

Heh.  They'll get over it. 😉

Perhaps this feature should be the opposite — the word "true" is not replaced with `true` **_unless_** the extra parameter is provided?  But that means all future code would need to provide said parameter.  (Again, an empty JSON object could signal the default mapping, but still...)

**Comment by Phergus** [comment]
**Macro 1**
```
[r: array = json.append("","truffles","true","false")]<br>
[r: array = json.set(array,2,"true")]<br>
[r: json.count(array,"true")]
```

**MapTool 1.5.4**
```
["truffles","true","false"]
["truffles","true","true"]
True Count: 2
```
**MapTool 1.5.5 beta 4**
```
["truffles",true,false]
["truffles",true,true]
True Count: 0
```
**Macro 2**
```
Original: [r: array = '["duck",false,"true","goose","frog"]']<br>
[r: "Set [0] True"]: [r: array = json.set(array,0,"true")]<br>
Count True:  [r: json.count(array,"true")]<br>
Count False: [r: json.count(array,"false")]<br>
[r: "Array[0]"]: [r: if(eval(json.get(array,0)),"It's true","It's not")]<br>
[r: "Array[1]"]: [r: if(eval(json.get(array,1)),"It's true","It's not")]
```
**MapTool 1.5.4**
```
Original: ["duck",false,"true","goose","frog"]
Set [0] True: ["true",false,"true","goose","frog"]
Count True: 2
Count False: 0
Array[0]: It's true
Array[1]: It's not
```
**MapTool 1.5.5 beta 4**
```
Original: ["duck",false,"true","goose","frog"]
Set [0] True: [true,false,"true","goose","frog"]
Count True: 1
Count False: 0
Array[0]: It's true
Array[1]: It's not
```

**Comment by Azhrei** [comment]
What is your point?

The goal was not to allow MT’s concept of true and false to be represented inside of JSON, but to allow the JSON to contain actual Boolean values when “true” and “false” were used.  It would be nice if MT’s true/false converted properly, but technically that’s a separate issue.  (I’m fine with rolling it into this issue, depending on how much work is involved.  It could be quite a bit.)

**Comment by Phergus** [comment]
No real point.   I didn't see anything resembling a consensus on this before the code was merged in.  Stuff worked one way before and now it's different.   Is this working correctly?  I have no idea.

Not my circus; not my monkeys.

**Comment by Merudo** [comment]
I'm in same boat of @Phergus - I don't really understand what PR #672 is trying to do exactly, so I have no idea if it is working properly or not,

**Comment by bobifle** [comment]
Quoting my original post:

If we implement the fix I proposed, then:

> We loose the ability to properly encode the actual strings "false", "true", "null"

Phergus's code illustrate this very well. The behavior was expecting to change.
That was known before the PR (at least by me :D).

the rational behind all this is that you gain the ability to encode the json primitive types true, false and null, which are widely used in json, and you loose the ability to encore a very small set of strings: specifically: "true", "false", and "null".

But for the very few people, encoding bools in json using strings "true", "false", I would point to:
https://xkcd.com/1172/ :)

However, that being said, Phergus's code made me realize there's an alternative solution that requires ... no change.
It seems with the current maptool that you can actually set true, false primitive types by encoding the json yourself in MT with a string.

with current 1.5.4:
```
<!-- Alternative to the currently invalid json.append(array, true)-->
[h: array = "[1,2,3,4,5]"]
[r: array = json.merge(array, "[true]")]

<!-- Alternative to the currently invalid json.set(obj, true)-->
[h: obj = '{"key0": 1}']
[r: obj = json.merge(obj, '{"key2": true}')]
output:   
[1,2,3,4,5,true]              
{"key0":1,"key2":true}
```
would yield a json array with an actual bool in the internal java object, resulting in the proper type sent to the remote, and you dont need PR #672 anymore.

```
<!-- but it's not perfect, types are not consistent whenever values are evaluated by MT-->
<!-- just re-setting key2 with it's own value
[r: obj = json.set(obj, "key2", json.get(obj, "key2"))]
{"key0":1,"key2":"true"}  see true is now a string :-/
```

Azhrei proposal to parse propely the true, false token in the MT parser woud be the best solution of course, but it's far beyond my java skills, so that would be without me :o)

**Comment by Phergus** [comment]
Tested.  Working as expected.

Closing issue.
